File: nexago_app/lib/features/arenas/domain/virtual_slot_generator.dart
import 'arena_slot.dart';
import 'slots_query.dart';

/// Gera slots “virtuais” quando não há documentos em [arenaSlots] ou para preencher lacunas.
///
/// Lê `slotDurationMinutes` e `availabilitySchedule` do documento da quadra quando existirem;
/// caso contrário usa **08:00–22:00** a cada **60 min** (ajustável).
abstract final class VirtualSlotGenerator {
  VirtualSlotGenerator._();

  static const _defaultDurationMin = 60;
  static const _defaultStartMin = 8 * 60;
  static const _defaultEndMin = 22 * 60;

  static const _weekdayKeys = [
    'monday',
    'tuesday',
    'wednesday',
    'thursday',
    'friday',
    'saturday',
    'sunday',
  ];

  /// Mescla: para cada janela virtual, se existir documento com mesmo horário, mantém o persistido.
  static List<ArenaSlot> merge(
    List<ArenaSlot> persisted,
    List<ArenaSlot> virtual,
  ) {
    final byTime = <String, ArenaSlot>{};
    for (final p in persisted) {
      byTime[_timeKey(p)] = p;
    }
    final out = <ArenaSlot>[];
    final usedKeys = <String>{};
    for (final v in virtual) {
      final k = _timeKey(v);
      if (byTime.containsKey(k)) {
        out.add(byTime[k]!);
        usedKeys.add(k);
      } else {
        final overlap = persisted.firstWhere(
          (p) => _overlaps(p, v) && (p.isBooked || p.isBlocked),
          orElse: () => v,
        );
        if (!identical(overlap, v)) {
          out.add(
            v.copyWith(
              rawStatus: overlap.rawStatus,
              bookingId: overlap.bookingId,
              bookingAthleteId: overlap.bookingAthleteId,
            ),
          );
          usedKeys.add(_timeKey(overlap));
        } else {
          out.add(v);
        }
      }
    }
    for (final p in persisted) {
      final k = _timeKey(p);
      if (!usedKeys.contains(k)) out.add(p);
    }
    out.sort((a, b) => a.startTime.compareTo(b.startTime));
    return out;
  }

  static String _timeKey(ArenaSlot s) => '${s.startTime}_${s.endTime}';

  static bool _overlaps(ArenaSlot a, ArenaSlot b) {
    final aStart = _parseHm(a.startTime);
    final aEnd = _parseHm(a.endTime);
    final bStart = _parseHm(b.startTime);
    final bEnd = _parseHm(b.endTime);
    if (aStart == null || aEnd == null || bStart == null || bEnd == null) {
      return false;
    }
    final normAEnd = aEnd <= aStart ? aEnd + (24 * 60) : aEnd;
    final normBEnd = bEnd <= bStart ? bEnd + (24 * 60) : bEnd;
    return aStart < normBEnd && bStart < normAEnd;
  }

  static List<ArenaSlot> build({
    required SlotsQuery query,
    required Map<String, dynamic>? courtData,
    required DateTime date,
  }) {
    final duration = _readDuration(courtData);
    final ranges = _readRanges(courtData, date.weekday);
    final day = DateTime(date.year, date.month, date.day);

    final slots = <ArenaSlot>[];
    for (final range in ranges) {
      var cursor = range.startMin;
      while (cursor + duration <= range.endMin) {
        final start = _fmt(cursor);
        final end = _fmt(cursor + duration);
        slots.add(
          ArenaSlot.virtual(
            arenaId: query.arenaId,
            courtId: query.courtId,
            date: day,
            startTime: start,
            endTime: end,
            priceReais: query.fallbackPriceReais,
          ),
        );
        cursor += duration;
      }
    }
    return slots;
  }

  static int _readDuration(Map<String, dynamic>? courtData) {
    final v = courtData?['slotDurationMinutes'];
    final n = v is num ? v.toInt() : _defaultDurationMin;
    if (n < 15 || n > 240) return _defaultDurationMin;
    return n;
  }

  /// Faixas em minutos desde meia-noite para o dia da semana (1=seg … 7=dom).
  static List<_MinRange> _readRanges(Map<String, dynamic>? courtData, int weekday) {
    final sched = courtData?['availabilitySchedule'];
    if (sched is Map) {
      final key = _weekdayKeys[weekday - 1];
      final dayEntry = sched[key] ?? sched['$weekday'] ?? sched[weekday];
      if (dayEntry != null) {
        if (_isExplicitlyClosed(dayEntry)) return [];
        if (dayEntry is List && dayEntry.isEmpty) return [];
        final parsed = _parseDayEntry(dayEntry);
        if (parsed.isNotEmpty) return parsed;
        return [];
      }
    }
    if (sched is List) {
      for (final item in sched) {
        if (item is Map) {
          final wd = item['weekday'];
          final match = wd == weekday ||
              wd == _weekdayKeys[weekday - 1] ||
              wd == '$weekday';
          if (match) {
            final parsed = _parseDayEntry(item['ranges'] ?? item['slots'] ?? item);
            if (parsed.isNotEmpty) return parsed;
          }
        }
      }
    }
    return [
      _MinRange(_defaultStartMin, _defaultEndMin),
    ];
  }

  static bool _isExplicitlyClosed(dynamic dayEntry) {
    if (dayEntry is String && dayEntry.trim().toLowerCase() == 'closed') {
      return true;
    }
    if (dayEntry is bool && dayEntry == false) return true;
    if (dayEntry is Map && dayEntry['closed'] == true) return true;
    return false;
  }

  static List<_MinRange> _parseDayEntry(dynamic dayEntry) {
    if (dayEntry == null) return [];
    if (dayEntry is List) {
      final out = <_MinRange>[];
      for (final e in dayEntry) {
        if (e is Map) {
          final start = e['start'] ?? e['from'] ?? e['open'];
          final end = e['end'] ?? e['to'] ?? e['close'];
          final a = _parseHm(start);
          final b = _parseHm(end);
          final range = _rangeFromStartEnd(a, b);
          if (range != null) out.add(range);
        }
      }
      return out;
    }
    if (dayEntry is Map) {
      final start = dayEntry['start'] ?? dayEntry['from'];
      final end = dayEntry['end'] ?? dayEntry['to'];
      final a = _parseHm(start);
      final b = _parseHm(end);
      final range = _rangeFromStartEnd(a, b);
      if (range != null) return [range];
    }
    return [];
  }

  /// `00:00` como fechamento = meia-noite **final** do dia (24:00), p.ex. 23:00–00:00.
  static _MinRange? _rangeFromStartEnd(int? startMin, int? endMin) {
    if (startMin == null || endMin == null) return null;
    var end = endMin;
    if (end == 0 && startMin > 0) {
      end = 24 * 60;
    }
    if (end > startMin) return _MinRange(startMin, end);
    return null;
  }

  static int? _parseHm(dynamic v) {
    if (v is String) {
      final p = v.trim().split(':');
      if (p.length >= 2) {
        final h = int.tryParse(p[0]) ?? 0;
        final m = int.tryParse(p[1]) ?? 0;
        return h * 60 + m.clamp(0, 59);
      }
    }
    if (v is int) return v;
    return null;
  }

  static String _fmt(int minutes) {
    final w = minutes % (24 * 60);
    final h = w ~/ 60;
    final m = w % 60;
    return '${h.toString().padLeft(2, '0')}:${m.toString().padLeft(2, '0')}';
  }
}

class _MinRange {
  const _MinRange(this.startMin, this.endMin);
  final int startMin;
  final int endMin;
}

File: frontend/shared/arena-discovery/virtual-slot-generator.ts
import { arenaSlotVirtual, type ArenaSlot } from './arena-slot';
import { timeToMinutes } from './arena-slot';
import { virtualSlotPricing } from './court-pricing';
import type { ArenaPromotion } from './arena-promotion';
import type { SlotsQuery } from './slots-query';

const DEFAULT_DURATION_MIN = 60;
const DEFAULT_START_MIN = 8 * 60;
const DEFAULT_END_MIN = 22 * 60;

const WEEKDAY_KEYS = [
  'monday',
  'tuesday',
  'wednesday',
  'thursday',
  'friday',
  'saturday',
  'sunday',
] as const;

interface MinRange {
  startMin: number;
  endMin: number;
}

export function readSlotDurationMinutes(
  courtData: Record<string, unknown> | null | undefined,
): number {
  const v = courtData?.['slotDurationMinutes'];
  const n = typeof v === 'number' ? v : DEFAULT_DURATION_MIN;
  if (n < 15 || n > 240) {
    return DEFAULT_DURATION_MIN;
  }
  return n;
}

export function mergeSlots(persisted: ArenaSlot[], virtual: ArenaSlot[]): ArenaSlot[] {
  const byTime = new Map<string, ArenaSlot>();
  for (const p of persisted) {
    byTime.set(timeKey(p), p);
  }
  const out: ArenaSlot[] = [];
  const usedKeys = new Set<string>();

  for (const v of virtual) {
    const k = timeKey(v);
    const p = byTime.get(k);
    if (p) {
      out.push(p);
      usedKeys.add(k);
      continue;
    }
    const overlap = persisted.find(
      (x) => overlaps(x, v) && x.rawStatus.toLowerCase() !== 'available',
    );
    if (overlap) {
      out.push({ ...v, rawStatus: overlap.rawStatus });
      usedKeys.add(timeKey(overlap));
    } else {
      out.push(v);
    }
  }

  for (const p of persisted) {
    if (!usedKeys.has(timeKey(p))) {
      out.push(p);
    }
  }

  out.sort((a, b) => a.startTime.localeCompare(b.startTime) || a.courtId.localeCompare(b.courtId));
  return out;
}

export function buildVirtualSlots(params: {
  query: SlotsQuery;
  courtData: Record<string, unknown> | null | undefined;
  date: Date;
  promotions: ArenaPromotion[];
}): ArenaSlot[] {
  const duration = readSlotDurationMinutes(params.courtData);
  const ranges = readRanges(params.courtData, isoWeekday(params.date));
  const day = new Date(params.date.getFullYear(), params.date.getMonth(), params.date.getDate());
  const slots: ArenaSlot[] = [];

  for (const range of ranges) {
    let cursor = range.startMin;
    while (cursor + duration <= range.endMin) {
      const start = fmt(cursor);
      const end = fmt(cursor + duration);
      const pricing = virtualSlotPricing({
        courtId: params.query.courtId,
        date: day,
        startTime: start,
        slotDurationMinutes: duration,
        courtData: params.courtData,
        arenaFallbackPerHourReais: params.query.arenaFallbackPricePerHourReais,
        promotions: params.promotions,
      });
      slots.push(
        arenaSlotVirtual({
          arenaId: params.query.arenaId,
          courtId: params.query.courtId,
          date: day,
          startTime: start,
          endTime: end,
          priceReais: pricing?.finalSlotPriceReais ?? null,
          basePriceReais: pricing?.baseSlotPriceReais ?? null,
          appliedPromotionId: pricing?.appliedPromotionId ?? null,
        }),
      );
      cursor += duration;
    }
  }
  return slots;
}

function isoWeekday(date: Date): number {
  const d = date.getDay();
  return d === 0 ? 7 : d;
}

function readRanges(
  courtData: Record<string, unknown> | null | undefined,
  weekday: number,
): MinRange[] {
  const sched = courtData?.['availabilitySchedule'];
  if (sched && typeof sched === 'object' && !Array.isArray(sched)) {
    const map = sched as Record<string, unknown>;
    const key = WEEKDAY_KEYS[weekday - 1];
    const dayEntry = map[key] ?? map[String(weekday)];
    if (dayEntry != null) {
      if (isExplicitlyClosed(dayEntry)) {
        return [];
      }
      if (Array.isArray(dayEntry) && dayEntry.length === 0) {
        return [];
      }
      const parsed = parseDayEntry(dayEntry);
      if (parsed.length > 0) {
        return parsed;
      }
      return [];
    }
  }
  if (Array.isArray(sched)) {
    for (const item of sched) {
      if (item && typeof item === 'object') {
        const row = item as Record<string, unknown>;
        const wd = row['weekday'];
        const match =
          wd === weekday || wd === WEEKDAY_KEYS[weekday - 1] || wd === String(weekday);
        if (match) {
          const parsed = parseDayEntry(row['ranges'] ?? row['slots'] ?? item);
          if (parsed.length > 0) {
            return parsed;
          }
        }
      }
    }
  }
  return [{ startMin: DEFAULT_START_MIN, endMin: DEFAULT_END_MIN }];
}

function isExplicitlyClosed(dayEntry: unknown): boolean {
  if (typeof dayEntry === 'string' && dayEntry.trim().toLowerCase() === 'closed') {
    return true;
  }
  if (dayEntry === false) {
    return true;
  }
  if (dayEntry && typeof dayEntry === 'object' && (dayEntry as Record<string, unknown>)['closed'] === true) {
    return true;
  }
  return false;
}

function parseDayEntry(dayEntry: unknown): MinRange[] {
  if (dayEntry == null) {
    return [];
  }
  if (Array.isArray(dayEntry)) {
    const out: MinRange[] = [];
    for (const e of dayEntry) {
      if (e && typeof e === 'object') {
        const row = e as Record<string, unknown>;
        const range = rangeFromStartEnd(
          parseHm(row['start'] ?? row['from'] ?? row['open']),
          parseHm(row['end'] ?? row['to'] ?? row['close']),
        );
        if (range) {
          out.push(range);
        }
      }
    }
    return out;
  }
  if (typeof dayEntry === 'object') {
    const row = dayEntry as Record<string, unknown>;
    const range = rangeFromStartEnd(
      parseHm(row['start'] ?? row['from']),
      parseHm(row['end'] ?? row['to']),
    );
    return range ? [range] : [];
  }
  return [];
}

function rangeFromStartEnd(startMin: number | null, endMin: number | null): MinRange | null {
  if (startMin == null || endMin == null) {
    return null;
  }
  let end = endMin;
  if (end === 0 && startMin > 0) {
    end = 24 * 60;
  }
  if (end > startMin) {
    return { startMin, endMin: end };
  }
  return null;
}

function parseHm(v: unknown): number | null {
  if (typeof v === 'string') {
    const p = v.trim().split(':');
    if (p.length >= 2) {
      const h = Number.parseInt(p[0] ?? '', 10);
      const m = Number.parseInt(p[1] ?? '', 10);
      if (Number.isFinite(h) && Number.isFinite(m)) {
        return h * 60 + Math.min(59, Math.max(0, m));
      }
    }
  }
  if (typeof v === 'number') {
    return v;
  }
  return null;
}

function fmt(minutes: number): string {
  const w = minutes % (24 * 60);
  const h = Math.floor(w / 60);
  const m = w % 60;
  return `${String(h).padStart(2, '0')}:${String(m).padStart(2, '0')}`;
}

function timeKey(s: ArenaSlot): string {
  return `${s.startTime}_${s.endTime}`;
}

function overlaps(a: ArenaSlot, b: ArenaSlot): boolean {
  const aStart = timeToMinutes(a.startTime);
  const aEnd = timeToMinutes(a.endTime);
  const bStart = timeToMinutes(b.startTime);
  const bEnd = timeToMinutes(b.endTime);
  const normAEnd = aEnd <= aStart ? aEnd + 24 * 60 : aEnd;
  const normBEnd = bEnd <= bStart ? bEnd + 24 * 60 : bEnd;
  return aStart < normBEnd && bStart < normAEnd;
}
